File: examples/csharp/Patterns/Command.cs
namespace DesignPatterns.Command
{
    public interface ICommand
    {
        void Execute();
        void Undo();
    }

    public class TextDocument
    {
        public string Content { get; set; } = "";
        public override string ToString() => $"\"{Content}\"";
    }

    public class InsertTextCommand : ICommand
    {
        private readonly TextDocument _doc;
        private readonly string _text;

        public InsertTextCommand(TextDocument doc, string text) { _doc = doc; _text = text; }
        public void Execute() { _doc.Content += _text; Console.WriteLine($"  Insert \"{_text}\" -> {_doc}"); }
        public void Undo() { _doc.Content = _doc.Content[..^_text.Length]; Console.WriteLine($"  Undo insert -> {_doc}"); }
    }

    public class DeleteTextCommand : ICommand
    {
        private readonly TextDocument _doc;
        private readonly int _count;
        private string _deleted = "";

        public DeleteTextCommand(TextDocument doc, int count) { _doc = doc; _count = count; }
        public void Execute()
        {
            _deleted = _doc.Content[^_count..];
            _doc.Content = _doc.Content[..^_count];
            Console.WriteLine($"  Delete {_count} chars -> {_doc}");
        }
        public void Undo() { _doc.Content += _deleted; Console.WriteLine($"  Undo delete -> {_doc}"); }
    }

    public class Editor
    {
        private readonly Stack<ICommand> _history = new();

        public void Execute(ICommand cmd)
        {
            cmd.Execute();
            _history.Push(cmd);
        }

        public void Undo()
        {
            if (_history.Count > 0)
                _history.Pop().Undo();
        }
    }

    public static class Demo
    {
        public static void Run()
        {
            var doc = new TextDocument();
            var editor = new Editor();

            editor.Execute(new InsertTextCommand(doc, "Hello"));
            editor.Execute(new InsertTextCommand(doc, " World"));
            editor.Execute(new DeleteTextCommand(doc, 5));
            editor.Undo();
            editor.Undo();
        }
    }
}

File: examples/cpp/command.cpp
#include <iostream>
#include <memory>
#include <stack>
#include <string>

class TextEditor {
    std::string text_;
public:
    void insertText(const std::string& t) { text_ += t; }
    void deleteText(size_t count) {
        if (count <= text_.size()) text_.erase(text_.size() - count);
    }
    std::string getText() const { return text_; }
    void setText(const std::string& t) { text_ = t; }
};

class Command {
public:
    virtual ~Command() = default;
    virtual void execute() = 0;
    virtual void undo() = 0;
};

class InsertCommand : public Command {
    TextEditor& editor_;
    std::string text_;
public:
    InsertCommand(TextEditor& ed, const std::string& text) : editor_(ed), text_(text) {}
    void execute() override { editor_.insertText(text_); }
    void undo() override { editor_.deleteText(text_.size()); }
};

class DeleteCommand : public Command {
    TextEditor& editor_;
    size_t count_;
    std::string deleted_;
public:
    DeleteCommand(TextEditor& ed, size_t count) : editor_(ed), count_(count) {}
    void execute() override {
        std::string t = editor_.getText();
        deleted_ = t.substr(t.size() - count_);
        editor_.deleteText(count_);
    }
    void undo() override { editor_.insertText(deleted_); }
};

class CommandInvoker {
    std::stack<std::unique_ptr<Command>> history_;
public:
    void executeCommand(std::unique_ptr<Command> cmd) {
        cmd->execute();
        history_.push(std::move(cmd));
    }
    void undo() {
        if (!history_.empty()) {
            history_.top()->undo();
            history_.pop();
        }
    }
};

int main() {
    std::cout << "=== Command: Text Editor Operations ===" << std::endl;

    TextEditor editor;
    CommandInvoker invoker;

    invoker.executeCommand(std::make_unique<InsertCommand>(editor, "Hello"));
    std::cout << "After insert: \"" << editor.getText() << "\"" << std::endl;

    invoker.executeCommand(std::make_unique<InsertCommand>(editor, ", World!"));
    std::cout << "After insert: \"" << editor.getText() << "\"" << std::endl;

    invoker.executeCommand(std::make_unique<DeleteCommand>(editor, 7));
    std::cout << "After delete: \"" << editor.getText() << "\"" << std::endl;

    invoker.undo();
    std::cout << "After undo:   \"" << editor.getText() << "\"" << std::endl;

    invoker.undo();
    std::cout << "After undo:   \"" << editor.getText() << "\"" << std::endl;

    return 0;
}
